### PR TITLE
feat: added Kanban Quick Card

### PR DESCRIPTION
**Pull request by bhattdevarsh**
[TASK](https://app.asana.com/0/1192407897953440/1195337199620453)

kanban

### CODE REVIEW

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 1]
```diff
@@ -822,3 +824,50 @@ frappe.provide("frappe.views");
...
 	}
 })();
+
+
+frappe.ui.kanbanForm = class {
+
+	constructor(doctype, docname){
+		console.log("Constructor");
+		if(!doctype || !docname){
+			frappe.throw(__("Pelase provide Document Type and/or Document Name"))
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]:
```suggestion
			frappe.throw(__("Document not found."));
```

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 2]
```diff
@@ -822,3 +824,50 @@ frappe.provide("frappe.views");
...
+			this.meta=frappe.get_meta(this.doctype)
+
+			this.meta.fields.forEach(field => {
+				if(field.fieldtype != "Table MultiSelects"){
+					this.fields.push(field)
+				}
+			});
+			this.mandatory = $.map(this.fields, function(d) {
+				return ((d.reqd || d.bold || d.allow_in_quick_entry) && !d.read_only) ? $.extend({}, d) : null;
+			});
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]: - Check if you can show all the fields in dialog.
- Also check if we can show table in dialog.

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 3]
```diff
@@ -822,3 +824,50 @@ frappe.provide("frappe.views");
...
+	set_meta_and_doc() {
+		this.fields= []
+		frappe.model.with_doctype(this.doctype,() =>{
+			this.meta=frappe.get_meta(this.doctype)
+
+			this.meta.fields.forEach(field => {
+				if(field.fieldtype != "Table MultiSelects"){
+					this.fields.push(field)
+				}
+			});
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]: - Show all the fields

**Review by hrwX — CHANGES_REQUESTED** [review]

**Comment by neilLasrado** [comment]
@bhattdevarsh its been 10 days since Himanshu gave a review. Please take this up on priority.

**Comment by bhattdevarsh** [comment]
> @bhattdevarsh its been 10 days since Himanshu gave a review. Please take this up on priority.

hey @neilLasrado, Me and himanshu have discussed on this. as of now i continue working on client request and bugs for that reason This feature building has been kept on halt.

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 4]
```diff
@@ -209,3 +209,18 @@ def save_filters(board_name, filters):
 	'''Save filters silently'''
 	frappe.db.set_value('Kanban Board', board_name, 'filters',
 						filters, update_modified=False)
+
+@frappe.whitelist()
+def on_submit_kanban_dialog(docname,doctype,values):
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
def on_submit_kanban_dialog(docname, doctype, values):
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 5]
```diff
@@ -209,3 +209,18 @@ def save_filters(board_name, filters):
...
+	'''To update and save kanban card dialog box'''
+	value = json.loads(values)
+	task_doc = frappe.get_doc(doctype,docname)
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	task_doc = frappe.get_doc(doctype, docname)
```

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 6]
```diff
@@ -822,3 +824,83 @@ frappe.provide("frappe.views");
...
+			this.meta.fields.forEach(field => {
+				if (field.fieldtype != "Table MultiSelects" && field.fieldtype != "Table") {
+					this.fields.push(field);
+				}
+			});
+			this.mandatory = $.map(this.fields, function (d) {
+				return ((d.reqd || d.bold || d.allow_in_quick_entry) && !d.read_only) ? $.extend({}, d) : null;
+			});
+			// Push Assignment Field For quick Entry
+			this.mandatory.push({ fieldtype: 'Link', fieldname: 'assign_to', options: 'User', label: __("Assign To"), reqd: false, filters: { 'user_type': 'System User' } });
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]:
```suggestion
			this.mandatory.push({ fieldtype: 'MultiSelectPills', fieldname: 'assign_to', options: 'User', label: __("Assign To"), reqd: false, filters: { 'user_type': 'System User' } });
```

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 7]
```diff
@@ -822,3 +824,83 @@ frappe.provide("frappe.views");
...
+			frappe.db.get_doc(this.doctype, this.docname).then(r => {
+				this.doc = r;
+				this.create_edit_dialog();
+			});
+		})
+	}
+
+	create_edit_dialog() {
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]:
```suggestion
	create_dialog() {
```

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 8]
```diff
@@ -822,3 +824,83 @@ frappe.provide("frappe.views");
...
+				if (values.assign_to) {
+					frappe.call({
+						method: "frappe.desk.form.assign_to.add", // call assign to and add functionality
+						args: {
+							'doctype': me.doctype,
+							'name': me.docname,
+							'assign_to': values.assign_to,
+						}
+					})
+				}
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]: - Call `assign_to.add` from `on_submit_kanban_dialog`, do not make a new `frappe.call`

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 9]
```diff
@@ -209,3 +210,26 @@ def save_filters(board_name, filters):
 	'''Save filters silently'''
 	frappe.db.set_value('Kanban Board', board_name, 'filters',
 						filters, update_modified=False)
+
+@frappe.whitelist()
+def on_submit_kanban_dialog(docname, doctype, values):
+	"""To update and save kanban card from dialog box !"""
+	value = json.loads(values)
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	values = json.loads(values)
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 10]
```diff
@@ -209,3 +210,26 @@ def save_filters(board_name, filters):
...
+	task_doc = frappe.get_doc(doctype, docname)
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	doc = frappe.get_doc(doctype, docname)
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 11]
```diff
@@ -209,3 +210,26 @@ def save_filters(board_name, filters):
...
+		if key in "assign_to":
+			for assignee in value[key]:
+				args = {
+					'doctype': doctype,
+					'name': docname,
+					'assign_to': assignee,
+				}
+				add(args)
+			continue
+		updates.update({key: value[key]})
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	assignees = values.pop("assign_to")
	add(assignees)
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 12]
```diff
@@ -209,3 +210,26 @@ def save_filters(board_name, filters):
...
 						filters, update_modified=False)
+
+@frappe.whitelist()
+def on_submit_kanban_dialog(docname, doctype, values):
+	"""To update and save kanban card from dialog box !"""
+	value = json.loads(values)
+	task_doc = frappe.get_doc(doctype, docname)
+	keys = value.keys()
+	updates = {}
+	args = {}
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 13]
```diff
@@ -209,3 +210,26 @@ def save_filters(board_name, filters):
...
+				args = {
+					'doctype': doctype,
+					'name': docname,
+					'assign_to': assignee,
+				}
+				add(args)
+			continue
+		updates.update({key: value[key]})
+	task_doc.update(updates)
+	task_doc.save()
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	doc.update(updates)
	doc.save()
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Comment by bhattdevarsh** [comment]
@hrwX , these suggestions seems sorted. but we need the keys to map from values to the doc.
and also we are removing `updates = {}` so now we cannot do `doc.update(updates)`

**Comment by hrwX** [comment]
> @hrwX , these suggestions seems sorted. but we need the keys to map from values to the doc.
> and also we are removing `updates = {}` so now we cannot do `doc.update(updates)`

The values are being sent in a dict.

File: frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py [hunk 14]
```diff
@@ -5,23 +5,23 @@
...
 def execute(filters=None):
 	user, doctype, show_permissions = filters.get("user"), filters.get("doctype"), filters.get("show_permissions")
+
 	if not validate(user, doctype): return [], []
 
 	columns, fields = get_columns_and_fields(doctype)
 	data = frappe.get_list(doctype, fields=fields, as_list=True, user=user)
 
 	if show_permissions:
-		columns = columns + ["Read", "Write", "Create", "Delete", "Submit", "Cancel", "Amend", "Print", "Email",
```
Comment (hrwX) [inline frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py]: Do not import rights from permission.py, revert this code change

File: frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py [hunk 15]
```diff
@@ -41,11 +41,13 @@ def get_columns_and_fields(doctype):
 
 	return columns, fields
 
+@frappe.whitelist()
+# @frappe.validate_and_sanitize_search_inputs
```
Comment (hrwX) [inline frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py]:
```suggestion
```

File: frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py [hunk 16]
```diff
@@ -41,11 +41,13 @@ def get_columns_and_fields(doctype):
...
 def query_doctypes(doctype, txt, searchfield, start, page_len, filters):
 	user = filters.get("user")
 	user_perms = frappe.utils.user.UserPermissions(user)
 	user_perms.build_permissions()
-	can_read = user_perms.can_read
+	can_read = user_perms.can_read # Does not include child tables
```
Comment (hrwX) [inline frappe/core/report/permitted_documents_for_user/permitted_documents_for_user.py]: Why are there changes to permitted_documents ?

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 17]
```diff
@@ -822,3 +824,82 @@ frappe.provide("frappe.views");
...
+		/**
+		 * Pull all Required, Bold, Allow Quick Entry Fields to show on Kanban Dialog
+		 */
+		this.fields = [];
+		frappe.model.with_doctype(this.doctype, () => {
+			this.meta = frappe.get_meta(this.doctype);
+
+			// Escape field type "Table MultiSelects" and "Table"
+			this.meta.fields.forEach(field => {
+				if (field.fieldtype != "Table MultiSelects" && field.fieldtype != "Table") {
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]: Can you try Table field too in dialog, since they support Tables in dialog

**Review by hrwX — DISMISSED** [review]

**Comment by hrwX** [comment]
@bhattdevarsh Please make the changes

File: frappe/public/js/frappe/views/kanban/kanban_board.js (hunk 17 above)
Comment (bhattdevarsh) [reply frappe/public/js/frappe/views/kanban/kanban_board.js]: as Discussed, its still showing the same error for table. so closing this comment

Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]:
```suggestion
				if (field.fieldtype != "Table MultiSelect" && field.fieldtype != "Table") {
```

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 18]
```diff
@@ -822,3 +824,82 @@ frappe.provide("frappe.views");
...
+					this.fields.push(field);
+				}
+			});
+			this.mandatory = $.map(this.fields, function (d) {
+				return ((d.reqd || d.bold || d.allow_in_quick_entry) && !d.read_only) ? $.extend({}, d) : null;
+			});
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js:854]: Can we push all the fields to dialog except mandatory ones

File: frappe/public/js/frappe/views/kanban/kanban_board.js [hunk 19]
```diff
@@ -822,3 +824,82 @@ frappe.provide("frappe.views");
...
+
+	}
+
+	create_dialog() {
+		/**
+		 * create Kanban Dialog and handle on_submit and
+		 */
+		let me = this;
+		let d = new frappe.ui.Dialog({
+			title: 'Enter details',
```
Comment (hrwX) [inline frappe/public/js/frappe/views/kanban/kanban_board.js]:
```suggestion
			title: __("{0}: {1}", [this.doctype, this.docname]),
```

File: frappe/desk/doctype/kanban_board/kanban_board.py [hunk 20]
```diff
@@ -209,3 +210,18 @@ def save_filters(board_name, filters):
...
+	"""To update and save kanban card from dialog box !"""
+	values = json.loads(values)
+	doc = frappe.get_doc(doctype, docname)
+	assignees = values.pop("assign_to")
+	args = {
+		"doctype": doctype,
+		"name": docname,
+		"assign_to": assignees
+	}
+	add(args)
```
Comment (hrwX) [inline frappe/desk/doctype/kanban_board/kanban_board.py]:
```suggestion
	add({
		"doctype": doctype,
		"name": docname,
		"assign_to": assignees
	})
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/public/js/frappe/views/kanban/kanban_board.js (hunk 18 above)
Comment (bhattdevarsh) [reply frappe/public/js/frappe/views/kanban/kanban_board.js:854]: @hrwX we cannot have all the fields here, as it contains table as well. and we may also lose the essence of Quick Entry. and also i tried applying all the fields which are coming through the doc. but it gave error for `df`

**Review by hrwX — APPROVED** [review]

**Pull request merged**